blog/2024-09-12-building-the-opentofu-registry.md

```diff
@@ -0,0 +1,129 @@
+image: /img/blog/building-the-opentofu-registry.png
+---
+
+Recently, we published the beta version of the [OpenTofu Registry Search](https://search.opentofu.org), a user interface that lets you search for and view the documentation of providers and modules in the OpenTofu Registry. With this important milestone reached, it is finally time to talk about how the OpenTofu Registry and Search work under the hood and the pitfalls of running a public registry.
+
+## The Registry API
+
+OpenTofu and its predecessor Terraform rely on provider binaries created by the community to interact with various APIs. There are currently over 4,000 such providers in the OpenTofu Registry, enabling a wide range of integrations from cloud providers to managing your GitHub account. Anything that has an API to create some kind of resource can be integrated with OpenTofu.
+
+Aside from providers, the community has also created well over 20,000 reusable modules that implement higher level functionality, such as provisioning an entire infrastructure with a cloud provider with only a few configuration options.
+
+Since these providers and modules are created by the community, OpenTofu needs to know where to download them and what versions are available. This is where the Registry comes into play: it holds the information about available providers and modules, download URLs, checksums, and GPG keys for integrity verification.
+
+### Step 1: Service Discovery
+
+For both providers and modules, OpenTofu first requests the `https://registry.opentofu.org/.well-known/terraform.json` file, which has the following content:
+
+```json
+{
+  "modules.v1": "/v1/modules/",
+  "providers.v1": "/v1/providers/"
+}
+```
+
+This file lists the endpoints OpenTofu should query for information about modules and providers. For private registries there is also a third endpoint named `login.v1`, providing information about an OAuth endpoint to use for authentication. If you are interested in the details, you can read more about this protocol [in the OpenTofu documentation](https://opentofu.org/docs/internals/login-protocol/).
+
+### Step 2: Version Listing
+
+With the endpoint information, OpenTofu can query the version listing endpoint for the desired provider or module.
+
+- For providers this endpoint would be `/v1/providers/NAMESPACE/NAME/versions` ([example](https://registry.opentofu.org/v1/providers/opentofu/random/versions)).
+- For modules it would be `/v1/modules/NAMESPACE/NAME/SYSTEM/versions` ([example](https://registry.opentofu.org/v1/modules/hashicorp/consul/aws/versions)).
+
+### Step 3: Download Info
+
+Based on the information received, OpenTofu can request the information about the specific version, operating system and architecture.
+
+- For providers, it would is located at `/v1/providers/NAMESPACE/NAME/VERSION/download/OS/ARCH` ([example]( https://registry.opentofu.org/v1/providers/opentofu/random/2.0.0/download/linux/amd64)).
```

Comment:
```suggestion
- For providers, it would be located at `/v1/providers/NAMESPACE/NAME/VERSION/download/OS/ARCH` ([example]( https://registry.opentofu.org/v1/providers/opentofu/random/2.0.0/download/linux/amd64)).
```